cocoa/src/toga_cocoa/constraints.py

```diff
@@ -24,69 +25,79 @@ def __init__(self, widget):
         self.left_constraint = None
         self.top_constraint = None
 
+    def __del__(self):
+        if self.width_constraint:
+            self.width_constraint.release()
+        if self.height_constraint:
```

Comment:
You're correct that a widget can only be associated with a single constraints object - but if a widget is deleted, we need to free the associated constraints object, which in turn means releasing the ObjC instances that have been explicitly retained. 

The (missing) `retain()` calls weren't causing a problem on macOS, but they were on iOS; however, that's probably more a case of accidentally *not* hitting the problem, rather than the problem not being there. 

Exercising this could be interesting... I'll see what I can do.

Reply:
Yeah - I can't work out a way to reliably get coverage here. The garbage collector of the ObjC runtime and the Python garbage collector both need to agree that a widget has been destroyed before `__del__` is invoked on an interface widget; and that then needs to cascade down to the impl, and then the constraints object. 

I've added a nocover to the two affected `__del__` methods for now.

Reply:
In that case, isn't there a possible leak if a widget is removed from its container and then added to another container? When it's added to the new container, new constraints objects are created without releasing the old ones. 

The simplest solution would be to allocate the constraints objects in the constructor and then reuse them for all containers in the widget's lifetime. Or if that isn't possible because they're tied to the container, maybe the content of `__del__` could be factored out into another method, and called from the `if value is None and self.container` branch of `container.setter`.

Reply:
Good catch on the leak when reparenting; I'll fix that case.

Agreed that creating the constraint objects once would be preferable - except that the container is an argument to the top and left constraints. We'd also need to manage adding and removing the constraint from the existing containers. At that point, the accounting associated with destroying and creating fresh constraints isn't that far removed from what we're already doing.

Refactoring the cleanup code into a common location is definitely a good idea though, and it cleans up the setter as well, so I've taken that approach.